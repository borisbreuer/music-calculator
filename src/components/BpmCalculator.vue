<script setup lang="ts">
import { ref, reactive, computed, Ref } from 'vue'

let abSwitch = 0
let popArray: number | null = null
let timeArr: Ref<number[]> = ref([])
const timeA: Ref<number | null> = ref(null)
const timeB: Ref<number | null> = ref(null)
const calculating = ref(false)
const bpm: Ref<string> = ref('120');

function tabTempoHandler() {
  if(abSwitch === 0){
    timeA.value = Date.now();
    abSwitch = 1;
  } else if(abSwitch === 1) {
    timeB.value = Date.now();
    abSwitch = 0;
  }
  if(abSwitch === 0 && timeA.value !== null && timeB.value !== null ) {
    timeArr.value.push((timeB.value - timeA.value))
    if(timeArr.value.length > 4) timeArr.value.shift()
    bpm.value = (60000 / (timeArr.value.reduce((a, c) => (c + a), 0) / timeArr.value.length)).toFixed(2)
    calculating.value = true
  }
  if(timeArr.value.length){
    if(popArray === null){
      const time = timeArr.value[timeArr.value.length - 1] * 4 > 2000 ? 2000 : timeArr.value[timeArr.value.length - 1] * 4
      popArray = window.setInterval(() => {
        timeArr.value.shift()
        if(!timeArr.value.length && popArray != null) {
          clearInterval(popArray)
          abSwitch = 0
          timeArr.value = []
          timeA.value = null;
          timeB.value = null;
          calculating.value = false;
          popArray = null
        }
      }, time)
    }
  } 
}

const quaterNoteInMs = computed((): number => {
  let bpmTemp = parseFloat(bpm.value);
  if(bpmTemp === 0 || bpmTemp === null || bpmTemp === undefined || isNaN(bpmTemp)) return 0
  return (60000 / bpmTemp)
})

function calculateMs(fraction: number): number {
  return (quaterNoteInMs.value * fraction);
}

const trippletFraction = ref(0.6666666666666667);
const dottedFraction = ref(1.5);

const numerator = 4;
const denominators = [1, 2, 4, 8, 16, 32, 64, 128];

const notesObjArr = denominators.map((d) => {
  return (
    { 
      fraction: (numerator / d), 
      value: `1/${d}`
    }
  )
})

const notes = reactive(notesObjArr)
</script>

<template>
  <v-container class="fill-height">
    <v-responsive class="d-flex fill-height pt-4">
      <v-row class="d-flex align-center justify-center pb-0">
        <v-col cols="auto" class="pa-0 ma-0">
          <v-btn
            @click="tabTempoHandler"
            @keydown.ctrl.space="tabTempoHandler"
            :color="calculating ? 'amber' : 'deeporange'"
            class="mb-2"
            data-testid="tabtempo"
          >
            Tab Tempo [Space]
          </v-btn>
        </v-col>
      </v-row>
      <v-row class="d-flex align-center justify-center pb-0">
        <v-col cols="auto" class="pa-0 ma-0">
          <v-text-field
            v-model="bpm"
            data-testid="bpm"
            density="compact"
            color="deeporange"
            label="BPM"
            type="number"
            inputmode="decimal"
            max="260"
            size="2"
            maxlength="5"
            autofocus
            class="mt-2"
            prepend-icon="mdi-chevron-left"
            @click:prepend="() => bpm--"
            append-icon="mdi-chevron-right"
            @click:append="() => bpm++"
          />
        </v-col>
      </v-row>
      <v-row class="d-flex align-center justify-center pb-0">
        <v-col cols="10" class="pa-0 ma-0">
          <v-table>
            <thead>
              <tr>
                <th width="16,5%" class="text-left">Devision</th>
                <th width="16,5%" class="text-left">Time (ms)</th>
                <th width="16,5%" class="text-left">Devision dotted</th>
                <th width="16,5%" class="text-left">Time (ms)</th>
                <th width="16,5%" class="text-left">Devision tripplet</th>
                <th width="16,5%" class="text-left">Time (ms)</th>
              </tr>
            </thead>
            <tbody>
              <tr v-for="(note, i) in notes" :key="i">
                <td>{{ note.value }}</td>
                <td :data-testid="`msn-${i}`">{{ $n(calculateMs(note.fraction), 'decimal') }} ms</td>
                <td class="left-line">{{ note.value }} .</td>
                <td :data-testid="`msd-${i}`">{{ $n(calculateMs(note.fraction * dottedFraction), 'decimal') }} ms</td>
                <td class="left-line">{{ note.value }} T</td>
                <td :data-testid="`mst-${i}`">{{ $n(calculateMs(note.fraction * trippletFraction), 'decimal') }} ms</td>
              </tr>
            </tbody>
          </v-table>
        </v-col>
      </v-row>
    </v-responsive>
  </v-container>
</template>

<style lang="scss">
.left-line {
  border-left: thin solid rgba(var(--v-border-color), var(--v-border-opacity));
}

input::-webkit-outer-spin-button,
input::-webkit-inner-spin-button {
    /* display: none; <- Crashes Chrome on hover */
    -webkit-appearance: none;
    margin: 0;
}

input[type=number] {
    -moz-appearance: textfield; /* Firefox */
    appearance: textfield;
}
</style>
