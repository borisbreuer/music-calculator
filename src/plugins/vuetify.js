/**
 * plugins/vuetify.js
 *
 * Framework documentation: https://vuetifyjs.com`
 */

// Styles
import '@mdi/font/css/materialdesignicons.css'
import 'vuetify/styles'

// Composables
import { createVuetify } from "vuetify";

// https://vuetifyjs.com/en/introduction/why-vuetify/#feature-guides
export default createVuetify({
  defaults: {
    VBtn: { variant: "outlined" },
    VTextField: { variant: "outlined" },
  },
  theme: {
    defaultTheme: "dark",
    themes: {
      dark: {
        colors: {
          primary: "#8BC34A",
          secondary: "#03DAC6",
          error: "#B00020",
          info: "#2196F3",
          success: "#4CAF50",
          warning: "#FB8C00",
          yellow: "#FFEB3B",
          amber: "#FFC107",
          orange: "#FF9800",
          deeporange: "#FF5722",
        },
      },
    },
  },
});
