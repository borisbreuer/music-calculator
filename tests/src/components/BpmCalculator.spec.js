import { createVuetify } from "vuetify";
import { describe, it, expect } from "vitest";
import { mount } from "@vue/test-utils";
import BpmCalculator from "@/components/BpmCalculator.vue";

describe("BPM", () => {
  it("test", async () => {
    const vuetify = createVuetify();
    // expect(BpmCalculator).toBeTruthy();
    expect(1 + 1).toBe(2);

    const wrapper = mount(BpmCalculator, {
      props: {},
      global: {
        plugins: [vuetify],
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
