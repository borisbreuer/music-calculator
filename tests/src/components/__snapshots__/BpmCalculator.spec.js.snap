// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BPM > test 1`] = `
"<div class=\\"v-container fill-height\\">
  <div class=\\"v-responsive d-flex fill-height pt-4\\">
    <div class=\\"v-responsive__sizer\\"></div>
    <!---->
    <div class=\\"v-responsive__content\\">
      <div class=\\"v-row d-flex align-center justify-center pb-0\\">
        <div class=\\"v-col v-col-auto pa-0 ma-0\\">
          <div class=\\"v-input v-input--horizontal v-input--density-compact v-input--dirty v-text-field mt-2\\">
            <div class=\\"v-input__prepend\\">
              <!----><i class=\\"mdi-chevron-left mdi v-icon notranslate v-theme--light v-icon--size-default v-icon--clickable\\" role=\\"button\\" aria-hidden=\\"false\\" aria-label=\\"BPM prepended action\\"></i>
            </div>
            <div class=\\"v-input__control\\">
              <div class=\\"v-field v-field--active v-field--dirty v-field--variant-filled v-theme--light\\" role=\\"textbox\\">
                <div class=\\"v-field__overlay\\"></div>
                <div class=\\"v-field__loader\\">
                  <div class=\\"v-progress-linear v-theme--light\\" style=\\"top: 0px; height: 0px; --v-progress-linear-height: 2px; left: 50%; transform: translateX(-50%);\\" role=\\"progressbar\\" aria-hidden=\\"true\\" aria-valuemin=\\"0\\" aria-valuemax=\\"100\\">
                    <!---->
                    <div class=\\"v-progress-linear__background bg-deeporange\\" style=\\"width: 100%;\\"></div>
                    <transition-stub>
                      <div class=\\"v-progress-linear__indeterminate\\">
                        <div class=\\"v-progress-linear__indeterminate long bg-deeporange\\"></div>
                        <div class=\\"v-progress-linear__indeterminate short bg-deeporange\\"></div>
                      </div>
                    </transition-stub>
                    <!---->
                  </div>
                </div>
                <!---->
                <div class=\\"v-field__field\\" data-no-activator=\\"\\"><label class=\\"v-label v-field-label v-field-label--floating\\" aria-hidden=\\"true\\" for=\\"input-0\\">
                    <!---->BPM
                  </label><label class=\\"v-label v-field-label\\" for=\\"input-0\\">
                    <!---->BPM
                  </label>
                  <!----><input size=\\"5\\" type=\\"number\\" id=\\"input-0\\" aria-describedby=\\"input-0-messages\\" maxlength=\\"5\\" class=\\"v-field__input\\">
                  <!---->
                </div>
                <!---->
                <!---->
                <div class=\\"v-field__outline\\">
                  <!---->
                  <!---->
                </div>
              </div>
            </div>
            <div class=\\"v-input__append\\"><i class=\\"mdi-chevron-right mdi v-icon notranslate v-theme--light v-icon--size-default v-icon--clickable\\" role=\\"button\\" aria-hidden=\\"false\\" aria-label=\\"BPM appended action\\"></i>
              <!---->
            </div>
            <div class=\\"v-input__details\\">
              <transition-group-stub tag=\\"div\\" class=\\"v-messages\\" role=\\"alert\\" aria-live=\\"polite\\" id=\\"input-0-messages\\">
                <!---->
              </transition-group-stub>
              <!---->
            </div>
          </div>
        </div>
      </div>
      <div class=\\"v-row d-flex align-center justify-center pb-0\\">
        <div class=\\"v-col v-col-10 pa-0 ma-0\\">
          <div class=\\"v-table v-theme--light v-table--density-default\\">
            <!---->
            <div class=\\"v-table__wrapper\\">
              <table>
                <thead>
                  <tr>
                    <th width=\\"16,5%\\" class=\\"text-left\\">Devision</th>
                    <th width=\\"16,5%\\" class=\\"text-left\\">Time (ms)</th>
                    <th width=\\"16,5%\\" class=\\"text-left\\">Devision dotted</th>
                    <th width=\\"16,5%\\" class=\\"text-left\\">Time (ms)</th>
                    <th width=\\"16,5%\\" class=\\"text-left\\">Devision tripplet</th>
                    <th width=\\"16,5%\\" class=\\"text-left\\">Time (ms)</th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td>1/1</td>
                    <td>4.000,00 ms</td>
                    <td class=\\"left-line\\">1/1 .</td>
                    <td>6.000,00 ms</td>
                    <td class=\\"left-line\\">1/1 T</td>
                    <td>2.666,67 ms</td>
                  </tr>
                  <tr>
                    <td>1/2</td>
                    <td>2.000,00 ms</td>
                    <td class=\\"left-line\\">1/2 .</td>
                    <td>3.000,00 ms</td>
                    <td class=\\"left-line\\">1/2 T</td>
                    <td>1.333,33 ms</td>
                  </tr>
                  <tr>
                    <td>1/4</td>
                    <td>1.000,00 ms</td>
                    <td class=\\"left-line\\">1/4 .</td>
                    <td>1.500,00 ms</td>
                    <td class=\\"left-line\\">1/4 T</td>
                    <td>666,67 ms</td>
                  </tr>
                  <tr>
                    <td>1/8</td>
                    <td>500,00 ms</td>
                    <td class=\\"left-line\\">1/8 .</td>
                    <td>750,00 ms</td>
                    <td class=\\"left-line\\">1/8 T</td>
                    <td>333,33 ms</td>
                  </tr>
                  <tr>
                    <td>1/16</td>
                    <td>250,00 ms</td>
                    <td class=\\"left-line\\">1/16 .</td>
                    <td>375,00 ms</td>
                    <td class=\\"left-line\\">1/16 T</td>
                    <td>166,67 ms</td>
                  </tr>
                  <tr>
                    <td>1/32</td>
                    <td>125,00 ms</td>
                    <td class=\\"left-line\\">1/32 .</td>
                    <td>187,50 ms</td>
                    <td class=\\"left-line\\">1/32 T</td>
                    <td>83,33 ms</td>
                  </tr>
                  <tr>
                    <td>1/64</td>
                    <td>62,50 ms</td>
                    <td class=\\"left-line\\">1/64 .</td>
                    <td>93,75 ms</td>
                    <td class=\\"left-line\\">1/64 T</td>
                    <td>41,67 ms</td>
                  </tr>
                  <tr>
                    <td>1/128</td>
                    <td>31,25 ms</td>
                    <td class=\\"left-line\\">1/128 .</td>
                    <td>46,88 ms</td>
                    <td class=\\"left-line\\">1/128 T</td>
                    <td>20,83 ms</td>
                  </tr>
                </tbody>
              </table>
            </div>
            <!---->
          </div>
        </div>
      </div>
    </div>
  </div>
</div>"
`;
